File: 01. Document Viewer/01. WinForms/VB/Main.vb
Imports System.Drawing.Printing
Imports Gnostice.Core.Viewer
Imports Gnostice.Core
Imports Gnostice.Core.DOM

Public Class Form1

    Dim textSearchResult As TextSearchResult
    Dim textSearchOptions As TextSearchOptions
    Private Sub tsOpen_Click(sender As System.Object, e As System.EventArgs) Handles tsOpen.Click
        If odOpenDoc.ShowDialog() = DialogResult.OK Then
            txtZoom.Text = "100%"
            txtGoToPage.TextBoxTextAlign = HorizontalAlignment.Center
            txtGoToPage.Text = "1"
            Try
                DocumentViewer1.PageLayout = New PageLayout()
                DocumentViewer1.LoadDocument(odOpenDoc.FileName, "")
                Me.Text = System.IO.Path.GetFileName(odOpenDoc.FileName) + " - Gnostice Document Studio .NET"
            Catch ex As Exception
                MessageBox.Show(ex.Message)
            End Try
        End If
    End Sub

    Private Sub btnPrint_Click(sender As System.Object, e As System.EventArgs) Handles btnPrint.Click
        If DocumentViewer1.IsDocumentLoaded Then
            Dim printerOptionsDlg As New frmPrinterOptions()
            printerOptionsDlg.ShowDialog()

            If printerOptionsDlg.Print Then
                Dim settings As New PrinterSettings()
                settings.FromPage = 2
                settings.ToPage = 3
                settings.PrintRange = printerOptionsDlg.PrintRange
                settings.Copies = 1
                settings.DefaultPageSettings.Landscape = printerOptionsDlg.Landscape

                settings.DefaultPageSettings.Margins.Left = 0
                settings.DefaultPageSettings.Margins.Top = 0
                settings.DefaultPageSettings.Margins.Right = 0
                settings.DefaultPageSettings.Margins.Bottom = 0

                Throw New NotImplementedException
                'DocumentViewer1.PrintDocument(settings, printerOptionsDlg.PageSizingOptions, printerOptionsDlg.PagePositioningOptions, True, "2-4")
            End If
        End If

    End Sub

    Private Sub btnFirstPage_Click(sender As System.Object, e As System.EventArgs) Handles btnFirstPage.Click
        If DocumentViewer1.IsDocumentLoaded Then
            DocumentViewer1.FirstPage()
        End If
    End Sub

    Private Sub btnPrvPage_Click(sender As System.Object, e As System.EventArgs) Handles btnPrvPage.Click
        If DocumentViewer1.IsDocumentLoaded Then
            DocumentViewer1.PreviousPage()
        End If
    End Sub

    Private Sub btnNxtPage_Click(sender As System.Object, e As System.EventArgs) Handles btnNxtPage.Click
        If DocumentViewer1.IsDocumentLoaded Then
            DocumentViewer1.NextPage()
        End If
    End Sub

    Private Sub btnLastPage_Click(sender As System.Object, e As System.EventArgs) Handles btnLastPage.Click
        If DocumentViewer1.IsDocumentLoaded Then
            DocumentViewer1.LastPage()
        End If
    End Sub

    Private Sub tsActSize_Click(sender As System.Object, e As System.EventArgs) Handles tsActSize.Click
        If DocumentViewer1.IsDocumentLoaded Then
            DocumentViewer1.Zoom.ZoomMode = ZoomMode.ActualSize
        End If
    End Sub

    Private Sub tsFitPage_Click(sender As System.Object, e As System.EventArgs) Handles tsFitPage.Click
        If DocumentViewer1.IsDocumentLoaded Then
            DocumentViewer1.Zoom.ZoomMode = ZoomMode.FitPage
        End If
    End Sub

    Private Sub tsFitWidth_Click(sender As System.Object, e As System.EventArgs) Handles tsFitWidth.Click
        If DocumentViewer1.IsDocumentLoaded Then
            DocumentViewer1.Zoom.ZoomMode = ZoomMode.FitWidth
        End If
    End Sub

    Private Sub tsZoomOut_Click(sender As System.Object, e As System.EventArgs) Handles tsZoomOut.Click
        If DocumentViewer1.IsDocumentLoaded Then
            DocumentViewer1.ZoomOut()
        End If
    End Sub

    Private Sub tsZoomIn_Click(sender As System.Object, e As System.EventArgs) Handles tsZoomIn.Click
        If DocumentViewer1.IsDocumentLoaded Then
            DocumentViewer1.ZoomIn()
        End If
    End Sub

    Private Sub txtGoToPage_KeyDown(sender As System.Object, e As System.Windows.Forms.KeyEventArgs) Handles txtGoToPage.KeyDown
        Select Case e.KeyCode

            Case Keys.NumPad0, Keys.NumPad1, Keys.NumPad2, Keys.NumPad3, Keys.NumPad4, Keys.NumPad5, _
             Keys.NumPad6, Keys.NumPad7, Keys.NumPad8, Keys.NumPad9, Keys.Delete, Keys.[End], _
             Keys.Home, Keys.Down, Keys.Up, Keys.Left, Keys.Right, Keys.Back
                e.SuppressKeyPress = False
                Exit Select

            Case Keys.D0, Keys.D1, Keys.D2, Keys.D3, Keys.D4, Keys.D5, _
             Keys.D6, Keys.D7, Keys.D8, Keys.D9
                If e.Shift Then
                    e.SuppressKeyPress = True
                Else
                    e.SuppressKeyPress = False
                End If
                Exit Select

            Case Keys.Enter
                e.SuppressKeyPress = False
                If txtGoToPage.Text.Trim() <> "" Then
                    DocumentViewer1.GoToPage(Convert.ToInt32(txtGoToPage.Text))
                End If
                Exit Select
            Case Else

                e.SuppressKeyPress = True
                Exit Select
        End Select

    End Sub

    Private Sub txtZoom_KeyDown(sender As System.Object, e As System.Windows.Forms.KeyEventArgs)
        Select Case e.KeyCode

            Case Keys.NumPad0, Keys.NumPad1, Keys.NumPad2, Keys.NumPad3, Keys.NumPad4, Keys.NumPad5,
             Keys.NumPad6, Keys.NumPad7, Keys.NumPad8, Keys.NumPad9, Keys.Delete, Keys.[End],
             Keys.Home, Keys.Down, Keys.Up, Keys.Left, Keys.Right, Keys.Back
                e.SuppressKeyPress = False
                Exit Select

            Case Keys.D0, Keys.D1, Keys.D2, Keys.D3, Keys.D4, Keys.D5,
             Keys.D6, Keys.D7, Keys.D8, Keys.D9
                If e.Shift Then
                    e.SuppressKeyPress = True
                Else
                    e.SuppressKeyPress = False
                End If
                Exit Select

            Case Keys.Enter
                e.SuppressKeyPress = False
                DocumentViewer1.Zoom.ZoomPercent = Double.Parse(txtZoom.Text)
                Exit Select
            Case Else

                e.SuppressKeyPress = True
                Exit Select
        End Select
    End Sub

    Private Sub DocumentViewer1_ZoomChanged(sender As System.Object, e As ZoomChangedEventArgs) Handles DocumentViewer1.ZoomChanged
        txtZoom.Text = e.ZoomPercent.ToString() + "%"
    End Sub


    Private Sub DocumentViewer1_PageChanged(sender As System.Object, e As PageChangedEventArgs) Handles DocumentViewer1.PageChanged
        txtGoToPage.Text = e.PageNumber.ToString()
    End Sub

    Public Sub New()

        ' This call is required by the designer.
        InitializeComponent()
        ' Add any initialization after the InitializeComponent() call.

    End Sub


    Private Sub DocumentViewer1_NeedPassword(sender As System.Object, e As NeedPasswordEventArgs) Handles DocumentViewer1.NeedPassword

        Dim passwordDlg As New frmPassword()
        passwordDlg.ShowDialog()

        e.Password = passwordDlg.Password
        e.Cancel = False
    End Sub


    Private Sub ToolStripButton1_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ToolStripButton1.Click
        DocumentViewer1.RotatePagesAntiClockwise90()
    End Sub

    Private Sub ToolStripButton2_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ToolStripButton2.Click
        DocumentViewer1.RotatePagesClockwise90()
    End Sub

    Private Sub DocumentViewer1_Load(sender As Object, e As EventArgs) Handles DocumentViewer1.Load

    End Sub

    Private Sub ZoomPercent_SelectedIndexChanged(sender As Object, e As EventArgs) Handles txtZoom.SelectedIndexChanged
        If (DocumentViewer1 IsNot Nothing) Then
            DocumentViewer1.Zoom.ZoomPercent = Double.Parse(Replace(txtZoom.Text, "%", ""))
        End If
    End Sub

    Private Sub ToolStripButton11_Click(sender As Object, e As EventArgs) Handles ToolStripButton11.Click
        If (DocumentViewer1 IsNot Nothing) Then
            DocumentViewer1.PageLayout.LayoutMode = PageLayoutMode.SinglePage
            DocumentViewer1.PageLayout.ScrollMode = ScrollMode.NonContinuous
        End If
    End Sub

    Private Sub ToolStripButton12_Click(sender As Object, e As EventArgs) Handles ToolStripButton12.Click
        If (DocumentViewer1 IsNot Nothing) Then
            DocumentViewer1.PageLayout.LayoutMode = PageLayoutMode.SinglePage
            DocumentViewer1.PageLayout.ScrollMode = ScrollMode.Continuous
        End If
    End Sub

    Private Sub ToolStripButton13_Click(sender As Object, e As EventArgs) Handles ToolStripButton13.Click
        If (DocumentViewer1 IsNot Nothing) Then
            DocumentViewer1.PageLayout.LayoutMode = PageLayoutMode.TwoPage
            DocumentViewer1.PageLayout.ScrollMode = ScrollMode.NonContinuous
        End If
    End Sub

    Private Sub ToolStripButton14_Click(sender As Object, e As EventArgs) Handles ToolStripButton14.Click
        If (DocumentViewer1 IsNot Nothing) Then
            DocumentViewer1.PageLayout.LayoutMode = PageLayoutMode.TwoPage
            DocumentViewer1.PageLayout.ScrollMode = ScrollMode.Continuous
        End If
    End Sub

    Private Sub ToolStripButton3_Click(sender As Object, e As EventArgs) Handles ToolStripButton3.Click
        DocumentViewer1.RotatePagesAntiClockwise90(DocumentViewer1.CurrentPage)
    End Sub

    Private Sub ToolStripButton4_Click(sender As Object, e As EventArgs) Handles ToolStripButton4.Click
        DocumentViewer1.RotatePagesClockwise90(DocumentViewer1.CurrentPage)
    End Sub

    Private Sub ToolStripButton5_Click(sender As Object, e As EventArgs) Handles ToolStripButton5.Click
        DocumentViewer1.MouseMode = DOM.CursorPreferences.TextSelection
    End Sub

    Private Sub ToolStripButton6_Click(sender As Object, e As EventArgs) Handles ToolStripButton6.Click
        DocumentViewer1.MouseMode = DOM.CursorPreferences.Pan
    End Sub

    Private Sub ToolStripButton7_Click(sender As Object, e As EventArgs) Handles ToolStripButton7.Click
        DocumentViewer1.MouseMode = DOM.CursorPreferences.Marquee
    End Sub

    Private Sub ToolStripButton8_Click(sender As Object, e As EventArgs) Handles ToolStripButton8.Click
        DocumentViewer1.MouseMode = DOM.CursorPreferences.AreaSelection
    End Sub

    Private Sub ToolStripTextBox1_TextChanged(sender As Object, e As EventArgs) Handles txtSearchText.TextChanged

    End Sub

    Private Sub ToolStripTextBox1_KeyDown(sender As Object, e As KeyEventArgs) Handles txtSearchText.KeyDown
        If (e.KeyCode = System.Windows.Forms.Keys.Enter) Then
            Dim searchText As String = txtSearchText.Text.Trim()
            textSearchResult = DocumentViewer1.FindNext(searchText, TextSearchMode.Literal, textSearchOptions, textSearchResult)
            ShowSearchedTextNotFound(textSearchResult)
            e.Handled = True
            e.SuppressKeyPress = True
        End If
    End Sub

    Private Sub ShowSearchedTextNotFound(textSearchResult As TextSearchResult)
        If textSearchResult Is Nothing Then
            MessageBox.Show("DocumentViewer has finished searching the document. No matches were found.")
        End If
    End Sub

    Private Sub ToolStripSplitButton1_ButtonClick(sender As Object, e As EventArgs) Handles ToolStripSplitButton1.ButtonClick
        If (ToolStripSplitButton1.DropDown.Visible = True) Then
            ToolStripSplitButton1.HideDropDown()
        Else
            ToolStripSplitButton1.ShowDropDown()
        End If
    End Sub

    Private Sub ToolStripButton9_Click(sender As Object, e As EventArgs) Handles ToolStripButton9.Click
        Dim searchText As String = txtSearchText.Text.Trim()
        textSearchResult = DocumentViewer1.FindNext(searchText, TextSearchMode.Literal, textSearchOptions, textSearchResult)
        ShowSearchedTextNotFound(textSearchResult)
    End Sub

    Private Sub ToolStripButton10_Click(sender As Object, e As EventArgs) Handles ToolStripButton10.Click
        Dim searchText As String = txtSearchText.Text.Trim()
        textSearchResult = DocumentViewer1.FindPrevious(searchText, TextSearchMode.Literal, textSearchOptions, textSearchResult)
        ShowSearchedTextNotFound(textSearchResult)
    End Sub

    Private Sub WholeWordToolStripMenuItem_CheckedChanged(sender As Object, e As EventArgs) Handles WholeWordToolStripMenuItem.CheckedChanged
        If (WholeWordToolStripMenuItem.Checked = True) Then
            If Not textSearchOptions.HasFlag(TextSearchOptions.WholeWord) Then
                textSearchOptions = textSearchOptions Or TextSearchOptions.WholeWord
            End If
        Else
            If textSearchOptions.HasFlag(TextSearchOptions.WholeWord) Then
                textSearchOptions = textSearchOptions And Not TextSearchOptions.WholeWord
            End If
        End If
    End Sub

    Private Sub CaseSensitiveToolStripMenuItem_CheckedChanged(sender As Object, e As EventArgs) Handles CaseSensitiveToolStripMenuItem.CheckedChanged
        If CaseSensitiveToolStripMenuItem.Checked Then

            If Not textSearchOptions.HasFlag(TextSearchOptions.CaseSensitive) Then
                textSearchOptions = textSearchOptions Or TextSearchOptions.CaseSensitive
            End If
        Else

            If textSearchOptions.HasFlag(TextSearchOptions.CaseSensitive) Then
                textSearchOptions = textSearchOptions And Not TextSearchOptions.CaseSensitive
            End If
        End If
    End Sub

    Private Sub WrapAroundToolStripMenuItem_CheckedChanged(sender As Object, e As EventArgs) Handles WrapAroundToolStripMenuItem.CheckedChanged
        If WrapAroundToolStripMenuItem.Checked Then

            If Not textSearchOptions.HasFlag(TextSearchOptions.WrapAround) Then
                textSearchOptions = textSearchOptions Or TextSearchOptions.WrapAround
            End If
        Else

            If textSearchOptions.HasFlag(TextSearchOptions.WrapAround) Then
                textSearchOptions = textSearchOptions And Not TextSearchOptions.WrapAround
            End If
        End If
    End Sub
End Class

File: 01. Document Viewer/01. WinForms/C#/Main.cs
using Gnostice.Documents;
using Gnostice.Controls.WinForms;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using Gnostice.Core.Viewer;
using Gnostice.Core;
using System.IO;
using Gnostice.PageManager;
using Gnostice.Core.DOM;

namespace WinForms_CS_Document_Viewer
{
    public partial class Main : Form
    {
        TextSearchResult textSearchResult;
        TextSearchOptions textSearchOptions;
        public Main()
        {
            InitializeComponent();
            DocumentViewer1.DocumentLoaded += DocumentViewer1_DocumentLoaded;
            DocumentViewer1.ZoomValuesUpdated += DocumentViewer1_ZoomListUpdated;
        }

        private void DocumentViewer1_DocumentLoaded(object sender, DocumentLoadedEventArgs e)
        {
            textSearchResult = null;
            txtSearchText.Text = "";


        }

        private void DocumentViewer1_ZoomListUpdated(object sender, ZoomValuesUpdatedEventArgs e)
        {
            cmbZoomPercent.SelectedIndexChanged -= cmbZoomPercent_SelectedIndexChanged;
            int index = cmbZoomPercent.ComboBox.SelectedIndex;
            cmbZoomPercent.ComboBox.DataSource = DocumentViewer1.GetZoomList();
            cmbZoomPercent.ComboBox.DisplayMember = e.DisplayMember;
            cmbZoomPercent.ComboBox.ValueMember = e.ValueMember;
            cmbZoomPercent.SelectedIndexChanged += cmbZoomPercent_SelectedIndexChanged;
            if (e.Index != -1)
                cmbZoomPercent.ComboBox.SelectedIndex = e.Index;
            cmbZoomPercent.Text = e.Zoom.ToString("#.##") + "%";
        }

        private int oldScrollingMode = -1;
        private string Filename = string.Empty;

        void DocumentViewer1_ZoomChanged(object sender, ZoomChangedEventArgs e)
        {
            cmbZoomPercent.Text = DocumentViewer1.Zoom.ZoomPercent.ToString("#.##") + "%";

            txtColumns.Text = DocumentViewer1.PageLayout.Columns.ToString();
        }

        private void DocumentViewer1_PageCountChanged(object sender, PageCountChangedEventArgs e)
        {
            lblPageCount.Text = e.PageCount.ToString();
        }

        void DocumentViewer1_PageChanged(object sender, PageChangedEventArgs e)
        {
            txtCurrentPage.Text = e.PageNumber.ToString();
        }

        private void pageLayoutOptionToolStripMenuItem_Click(object sender, EventArgs e)
        {
            if (toolStripPageLayoutOptions.Visible == false)
            {
                toolStripPageLayoutOptions.Visible = true;
                pageLayoutOptionToolStripMenuItem.Image = WinForms_CS_Document_Viewer.Properties.Resources.tick;
            }
            else
            {
                toolStripPageLayoutOptions.Visible = false;
                pageLayoutOptionToolStripMenuItem.Image = null;
            }
        }

        private void btnShowCoverPage_Click(object sender, EventArgs e)
        {
            if (btnShowCoverPage.Image == null)
            {
                DocumentViewer1.PageLayout.ShowCoverPage = true;
                showCoverPageToolStripMenuItem.Image = WinForms_CS_Document_Viewer.Properties.Resources.tick;
                btnShowCoverPage.Image = WinForms_CS_Document_Viewer.Properties.Resources.tick;
            }
            else
            {
                DocumentViewer1.PageLayout.ShowCoverPage = false;
                showCoverPageToolStripMenuItem.Image = null;
                btnShowCoverPage.Image = null;
            }
        }

        private void navigationPaneToolStripMenuItem_Click(object sender, EventArgs e)
        {
            if (DocumentViewer1.NavigationPane.Visibility==Visibility.Auto)
            {
                navigationPaneToolStripMenuItem.Image = null;
                DocumentViewer1.NavigationPane.Visibility = Visibility.Never;
            }
            else
            {
                navigationPaneToolStripMenuItem.Image = WinForms_CS_Document_Viewer.Properties.Resources.tick;
                DocumentViewer1.NavigationPane.Visibility = Visibility.Auto;
                DocumentViewer1.Zoom.ZoomMode = ZoomMode.FitWidth;
            }
        }

        private void btnOpen_Click(object sender, EventArgs e)
        {
            if (openFileDialog1.ShowDialog() == System.Windows.Forms.DialogResult.OK)
            {
                try
                {
                    Filename = openFileDialog1.FileName;
                    this.Text = System.IO.Path.GetFileName(Filename) + " - Gnostice WinForms Document Viewer";
                    DocumentViewer1.LoadDocument(openFileDialog1.FileName);
                    cmbZoomPercent.FlatStyle = FlatStyle.Standard;

                    

                    //DocumentViewer1.Zoom.ZoomPercent = double.Parse(cmbZoomPercent.Text.Remove(cmbZoomPercent.Text.Length - 1));
                    DocumentViewer1_LastPageRead(DocumentViewer1, e as Gnostice.Core.LoadCompletedEventArgs);
                    lblPageCount.Text = DocumentViewer1.PageCount.ToString();
                }
                catch (Exception ex)
                {
                    MessageBox.Show(ex.Message);
                }
            }
        }

        private void DocumentViewer1_DragEnter(object sender, DragEventArgs e)
        {
            if (e.Data.GetDataPresent(DataFormats.FileDrop)) e.Effect = DragDropEffects.Copy;
        }

        private void DocumentViewer1_DragDrop(object sender, DragEventArgs e)
        {
            string[] files = (string[])e.Data.GetData(DataFormats.FileDrop);

            if (files.Length > 0)
            {
                try
                {
                    DocumentViewer1.LoadDocument(files[0]);

                    cmbZoomPercent.FlatStyle = FlatStyle.Standard;
                    DocumentViewer1.Zoom.ZoomPercent = double.Parse(cmbZoomPercent.Text.Remove(cmbZoomPercent.Text.Length - 1));
                    lblPageCount.Text = DocumentViewer1.PageCount.ToString();
                    DocumentViewer1_LastPageRead(DocumentViewer1, new Gnostice.Core.LoadCompletedEventArgs(DocumentViewer1.PageCount));
                    this.Filename = files[0];
                }
                catch (Exception ex)
                {
                    MessageBox.Show(ex.Message);
                }
            }
        }


        private void btnPrint_Click(object sender, EventArgs e)
        {
            if (DocumentViewer1.IsDocumentLoaded)
            {
                Print print = new Print(DocumentViewer1);
                print.ShowDialog();
            }
        }

        private void btnSave_Click(object sender, EventArgs e)
        {
            if (DocumentViewer1.IsDocumentLoaded)
            {
                string outputFormat = string.Empty;

                if (saveFileDialog1.ShowDialog() == System.Windows.Forms.DialogResult.OK)
                {
                    SavingOptions ep = new SavingOptions(Filename, saveFileDialog1.FileName);
                    ep.ShowDialog();
                }
            }
        }

        private void exitToolStripMenuItem_Click(object sender, EventArgs e)
        {
            this.Close();
        }

        private void btnFirst_Click(object sender, EventArgs e)
        {
            DocumentViewer1.FirstPage();
        }

        private void btnPrevious_Click(object sender, EventArgs e)
        {
            DocumentViewer1.PreviousPage();
        }

        private void btnNext_Click(object sender, EventArgs e)
        {
            DocumentViewer1.NextPage();
        }

        private void btnLast_Click(object sender, EventArgs e)
        {
            DocumentViewer1.LastPage();
        }

        private void btnZoomOut_Click(object sender, EventArgs e)
        {
            DocumentViewer1.ZoomOut();
            cmbZoomPercent.Text = DocumentViewer1.Zoom.ZoomPercent.ToString("#.##") + "%";
        }

        private void btnZoomIn_Click(object sender, EventArgs e)
        {
            DocumentViewer1.ZoomIn();
            cmbZoomPercent.Text = DocumentViewer1.Zoom.ZoomPercent.ToString("#.##") + "%";
        }

        private void btnRotateClockwise_Click(object sender, EventArgs e)
        {
            DocumentViewer1.RotatePagesClockwise90();
        }

        private void btnRotateAnticlockwise_Click(object sender, EventArgs e)
        {
            DocumentViewer1.RotatePagesAntiClockwise90();
        }

        private void btnRotateCurrentPageClockwise_Click(object sender, EventArgs e)
        {
            DocumentViewer1.RotatePagesClockwise90(DocumentViewer1.CurrentPage.ToString());
        }

        private void btnRotateCurrentPageAnticlockwise_Click(object sender, EventArgs e)
        {
            DocumentViewer1.RotatePagesAntiClockwise90(DocumentViewer1.CurrentPage.ToString());
        }

        private void btnSingleNonContinuous_Click(object sender, EventArgs e)
        {
            cmbLayoutMode.SelectedIndex = 0;
            cmbScrollMode.SelectedIndex = 1;
        }

        private void btnSingleContinuous_Click(object sender, EventArgs e)
        {
            cmbLayoutMode.SelectedIndex = 0;
            cmbScrollMode.SelectedIndex = 0;
        }

        private void btnTwoPageNonContinuous_Click(object sender, EventArgs e)
        {
            cmbLayoutMode.SelectedIndex = 1;
            cmbScrollMode.SelectedIndex = 1;
        }

        private void btnTwoPageContinuous_Click(object sender, EventArgs e)
        {
            cmbLayoutMode.SelectedIndex = 1;
            cmbScrollMode.SelectedIndex = 0;
        }

        private void allowPageShiftingToolStripMenuItem_Click(object sender, EventArgs e)
        {
            if (allowPageShiftingToolStripMenuItem.Image == null)
            {
                cmbScrollMode.SelectedIndex = oldScrollingMode;
                allowPageShiftingToolStripMenuItem.Image = WinForms_CS_Document_Viewer.Properties.Resources.tick;
            }
            else
            {
                oldScrollingMode = cmbScrollMode.SelectedIndex;
                cmbScrollMode.SelectedIndex = 2;
                allowPageShiftingToolStripMenuItem.Image = null;
            }
        }

        private void cmbLayoutMode_SelectedIndexChanged(object sender, EventArgs e)
        {
            switch (cmbLayoutMode.SelectedIndex)
            {
                case 0:
                    DocumentViewer1.PageLayout.LayoutMode = Gnostice.Core.Viewer.PageLayoutMode.SinglePage;
                    break;
                case 1:
                    DocumentViewer1.PageLayout.LayoutMode = Gnostice.Core.Viewer.PageLayoutMode.TwoPage;
                    break;
                case 2:
                    DocumentViewer1.PageLayout.LayoutMode = Gnostice.Core.Viewer.PageLayoutMode.AutoFitPagesInWindow;
                    break;
                case 3:
                    DocumentViewer1.PageLayout.LayoutMode = Gnostice.Core.Viewer.PageLayoutMode.SpecifiedColumns;
                    break;
            }

            txtColumns.Text = DocumentViewer1.PageLayout.Columns.ToString();
        }

        private void txtColumns_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == System.Windows.Forms.Keys.Enter)
            {
                try
                {
                    DocumentViewer1.PageLayout.Columns = int.Parse(txtColumns.Text);
                }
                catch { }
            }
        }

        private void cmbScrollMode_SelectedIndexChanged(object sender, EventArgs e)
        {
            switch (cmbScrollMode.SelectedIndex)
            {
                case 0:
                    DocumentViewer1.PageLayout.ScrollMode = Gnostice.Core.Viewer.ScrollMode.Continuous;
                    allowPageShiftingToolStripMenuItem.Image = WinForms_CS_Document_Viewer.Properties.Resources.tick;
                    break;
                case 1:
                    DocumentViewer1.PageLayout.ScrollMode = Gnostice.Core.Viewer.ScrollMode.NonContinuous;
                    allowPageShiftingToolStripMenuItem.Image = WinForms_CS_Document_Viewer.Properties.Resources.tick;
                    break;
                case 2:
                    DocumentViewer1.PageLayout.ScrollMode = Gnostice.Core.Viewer.ScrollMode.NonPageSwitching;
                    allowPageShiftingToolStripMenuItem.Image = null;
                    break;
            }

            txtColumns.Text = DocumentViewer1.PageLayout.Columns.ToString();
        }

        private void txtHorizontalSpacing_KeyDown(object sender, KeyEventArgs e)
        {
            //if (e.KeyCode == Keys.Enter)
            //{
            //    try
            //    {
            //        DocumentViewer1..PageLayout.HorizontalSpacing = int.Parse(txtHorizontalSpacing.Text);
            //    }
            //    catch { }
            //}
        }

        private void txtVerticalSpacing_KeyDown(object sender, KeyEventArgs e)
        {
            //if (e.KeyCode == Keys.Enter)
            //{
            //    try
            //    {
            //        DocumentViewer1..PageLayout.VerticalSpacing = int.Parse(txtVerticalSpacing.Text);
            //    }
            //    catch { }
            //}
        }

        private void shortcutManagerToolStripMenuItem_Click(object sender, EventArgs e)
        {
            ShortcutManager sm = new ShortcutManager();
            sm.ShowDialog();
        }

        private void cmbZoomPercent_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == System.Windows.Forms.Keys.Enter || e.KeyCode == System.Windows.Forms.Keys.Tab)
            {
                try
                {
                    DocumentViewer1.Zoom.ZoomPercent = double.Parse(cmbZoomPercent.Text.Split('%')[0]);
                }
                catch { }
            }
            e.Handled = true;
        }

        private void cmbZoomPercent_SelectedIndexChanged(object sender, EventArgs e)
        {
            if (DocumentViewer1 != null && DocumentViewer1.IsDocumentLoaded && cmbZoomPercent.SelectedIndex != -1)
            {
                ZoomInfo currentZoom = cmbZoomPercent.SelectedItem as ZoomInfo;
                if(currentZoom.ZoomType == ZoomType.Standard)
                {
                    DocumentViewer1.Zoom.ZoomMode = currentZoom.ZoomMode;
                }
                else
                {
                    DocumentViewer1.Zoom.ZoomPercent = (cmbZoomPercent.SelectedItem as ZoomInfo).Zoom;
                }
            }
        }

        private void DocumentViewer1_LastPageRead(object sender, Gnostice.Core.LoadCompletedEventArgs e)
        {
            cmbLayoutMode.SelectedIndex = 0;
            cmbScrollMode.SelectedIndex = 0;
            try
            {
                if (this.InvokeRequired)
                    this.Invoke(new Action(() => lblPageCount.Text = DocumentViewer1.PageCount.ToString()));
                if (this.InvokeRequired)
                    this.Invoke(new Action(() => cmbZoomPercent.Text = DocumentViewer1.Zoom.ZoomPercent.ToString("#.##") + "%"));
                else
                    cmbZoomPercent.Text = DocumentViewer1.Zoom.ZoomPercent.ToString("#.##") + "%";
                if (this.InvokeRequired)
                    this.Invoke(new Action(() => txtCurrentPage.Text = DocumentViewer1.CurrentPage.ToString()));
                else
                    txtCurrentPage.Text = DocumentViewer1.CurrentPage.ToString();
            }
            catch { }
        }


        
        private void Main_Resize(object sender, EventArgs e)
        {
            //if (DocumentViewer1 != null && DocumentViewer1..PageLayout != null)
            //    txtColumns.Text = DocumentViewer1..PageLayout.Columns.ToString();
        }

        private void btnArrow_Click(object sender, EventArgs e)
        {
            this.DocumentViewer1.MouseMode = Gnostice.Core.DOM.CursorPreferences.TextSelection;
        }

        private void btnPan_Click(object sender, EventArgs e)
        {
            this.DocumentViewer1.MouseMode = Gnostice.Core.DOM.CursorPreferences.Pan;
        }

        private void btnMarquee_Click(object sender, EventArgs e)
        {
            this.DocumentViewer1.MouseMode = Gnostice.Core.DOM.CursorPreferences.Marquee;
        }

        private void btnHorizontal_Click(object sender, EventArgs e)
        {
            DocumentViewer1.Orientation = Gnostice.Core.Viewer.ViewerOrientation.Horizontal;
        }

        private void btnVertical_Click(object sender, EventArgs e)
        {
            DocumentViewer1.Orientation = Gnostice.Core.Viewer.ViewerOrientation.Vertical;
        }

        float x1, y1, x2, y2, x, y;

        private void Main_Load(object sender, EventArgs e)
        {

        }

        private void AreaSelectionButton1_Click(object sender, EventArgs e)
        {
            this.DocumentViewer1.MouseMode = Gnostice.Core.DOM.CursorPreferences.AreaSelection;

        }

        private void txtSearchText_Validated(object sender, EventArgs e)
        {

        }

        private void txtSearchText_TextChanged(object sender, EventArgs e)
        {
            textSearchResult = null;
        }

        private void txtSearchText_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == System.Windows.Forms.Keys.Enter)
            {
                string searchText = txtSearchText.Text.Trim();
                textSearchResult = DocumentViewer1.FindNext(searchText, lastSearchResult: textSearchResult, searchOptions: textSearchOptions);
                ShowSearchedTextNotFound(textSearchResult);
                e.Handled = true;
                e.SuppressKeyPress = true;
            }
        }

        private void ShowSearchedTextNotFound(TextSearchResult result)
        {
            if (result == null)
            {
                MessageBox.Show("DocumentViewer has finished searching the document. No matches were found.");
            }
        }

        private void btnNextSearch_Click(object sender, EventArgs e)
        {
            string searchText = txtSearchText.Text.Trim();
            textSearchResult = DocumentViewer1.FindNext(searchText, lastSearchResult: textSearchResult, searchOptions: textSearchOptions);
            ShowSearchedTextNotFound(textSearchResult);
           
        }

        private void btnPrevSearch_Click(object sender, EventArgs e)
        {
            string searchText = txtSearchText.Text.Trim();
            textSearchResult = DocumentViewer1.FindPrevious(searchText, lastSearchResult: textSearchResult, searchOptions: textSearchOptions);
            ShowSearchedTextNotFound(textSearchResult);
        }


        private void wholeWorldToolStripMenuItem_CheckedChanged(object sender, EventArgs e)
        {
            if (wholeWorldToolStripMenuItem.Checked)
            {
                if (!textSearchOptions.HasFlag(TextSearchOptions.WholeWord))
                {
                    textSearchOptions |= TextSearchOptions.WholeWord;
                }
            }
            else
            {
                if (textSearchOptions.HasFlag(TextSearchOptions.WholeWord))
                {
                    textSearchOptions &= ~TextSearchOptions.WholeWord;
                }
            }
        }

        private void caseSensitiveToolStripMenuItem_CheckedChanged(object sender, EventArgs e)
        {
            if (caseSensitiveToolStripMenuItem.Checked)
            {
                if (!textSearchOptions.HasFlag(TextSearchOptions.CaseSensitive))
                {
                    textSearchOptions |= TextSearchOptions.CaseSensitive;
                }
            }
            else
            {
                if (textSearchOptions.HasFlag(TextSearchOptions.CaseSensitive))
                {
                    textSearchOptions &= ~TextSearchOptions.CaseSensitive;
                }
            }
        }

        private void wrapAroundToolStripMenuItem_CheckedChanged(object sender, EventArgs e)
        {
            if (wrapAroundToolStripMenuItem.Checked)
            {
                if (!textSearchOptions.HasFlag(TextSearchOptions.WrapAround))
                {
                    textSearchOptions |= TextSearchOptions.WrapAround;
                }
            }
            else
            {
                if (textSearchOptions.HasFlag(TextSearchOptions.WrapAround))
                {
                    textSearchOptions &= ~TextSearchOptions.WrapAround;
                }
            }
        }

        private void toolStripSplitButton1_ButtonClick(object sender, EventArgs e)
        {
            if (toolStripSplitButton1.DropDown.Visible)
                toolStripSplitButton1.HideDropDown();
            else
                toolStripSplitButton1.ShowDropDown();
        }

        private void TxtCurrentPage_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == System.Windows.Forms.Keys.Enter || e.KeyCode == System.Windows.Forms.Keys.Tab)
            {
                int pageNumber;
                if (int.TryParse(txtCurrentPage.Text, out pageNumber))
                    DocumentViewer1.GoToPage(pageNumber);
                else
                    txtCurrentPage.Text = DocumentViewer1.CurrentPage.ToString();
            }
        }

        Stream imageStream;

        private void DocumentViewer1_MouseUp(object sender, Gnostice.Core.Viewer.MouseActivityEventArgs e)
        {
            x2 = (e as WinFormsMouseActivityEventArgs).PosRelativetoPage.X;
            y2 = (e as WinFormsMouseActivityEventArgs).PosRelativetoPage.Y;
            Rectangle rect = new Rectangle((int)Math.Min(x1, x2), (int)Math.Min(y1, y2), (int)Math.Abs(x1 - x2), (int)Math.Abs(y1 - y2));
            imageStream = e.ImageStream;
        }

        private void DocumentViewer1_MouseDown(object sender, Gnostice.Core.Viewer.MouseActivityEventArgs e)
        {
            x1 = (e as WinFormsMouseActivityEventArgs).PosRelativetoPage.X;
            y1 = (e as WinFormsMouseActivityEventArgs).PosRelativetoPage.Y;
        }

        private void DocumentViewer1_MouseMove(object sender, Gnostice.Core.Viewer.MouseActivityEventArgs e)
        {
            x = (e as WinFormsMouseActivityEventArgs).PosRelativetoPage.X;
            y = (e as WinFormsMouseActivityEventArgs).PosRelativetoPage.Y;
        }
        
    }
}
